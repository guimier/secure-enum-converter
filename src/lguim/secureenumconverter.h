#ifndef LGUIM_SECURE_ENUM_CONVERTER
#define LGUIM_SECURE_ENUM_CONVERTER

#include <set>
#include <stdexcept>
#include <type_traits>

#ifndef SEC_OPTIONAL_NS
#define SEC_OPTIONAL_NS std
#include <optional>
#endif

namespace lguim {

/** `SecureEnumConverter` is a bi-directional enum converter, which
 * needs only one mapping in code to do both directions, and will
 * only compile when the behavior for all values is explicitely
 * written in code.
 *
 * @tparam InternalType One of the two enumeration types.
 * @tparam ExternalType The other enumeration type.
 * @tparam Tag This optional type can be defined to any custom type
 *     when you need to have two different mappings between the same
 *     types. There is no need for the type to be instanciable, only
 *     to have an identity.
 *     Exemple:
 *     ````
 *     enum class A { A1, A2 };
 *     enum class B { B1, B2 };
 *     using StandardConverter = SecureEnumConverter<A, B, struct StandardConverterTag>;
 *     using AlternateConverter = SecureEnumConverter<A, B, struct AlternateConverterTag>;
 *     ```
 *
 * The internal/external naming is used everywhere in the interface
 * to differenciate the direction. Finding a good naming being at
 * the same time generic, easy to use and easy to read is hard, and
 * we believe the internal/external distinction is the one that will
 * be most helpful where this class is needed. We apologize if this
 * is not the case for you.
 *
 * The implementation of the methods should sit in its own translation
 * unit.
 *
 * To give this class the correct mapping to use, you need to define
 * two macros and include a file:
 *
 * ```
 * enum class A { A1, A2 };
 * enum class B { B1, B2 };
 * using Converter = lguim::SecureEnumConverter<A, B>;
 *
 * #define SEC_TYPE Converter
 * #define SEC_MAPPING \
 *     SEC_EQUIV(A::A1, B::B1) \
 *     SEC_EQUIV(A::A2, B::B2)
 * #include "lguim/secureenumconverter.inc"
 * ```
 *
 * This inclusion must happen outside any namespace. It can happen
 * multiple times in the same translation unit.
 *
 * The following macros are available for the mapping:
 *
 *   - `SEC_EQUIV` describes a direct equivalence between two values
 *     SEC_EQUIV(x_int, x_ext) ⇔ { to_ext(x_int) = x_ext; to_int(x_ext) = x_int }
 *   - `SEC_PROJ_I2E` describes a unidirectional conversion from internal to external
 *     SEC_PROJ_I2E(x_int, x_ext) ⇔ { to_ext(x_int) = x_ext; to_int(x_ext) is defined separately }
 *   - `SEC_PROJ_E2I` describes a unidirectional conversion from external to internal
 *     SEC_PROJ_E2I(x_int, x_ext) ⇔ { to_int(x_ext) = x_int; to_ext(x_int) is defined separately }
 *   - `SEC_ORPHAN_INT` marks an internal value as having no matching external value
 *     SEC_ORPHAN_INT(x_int) ⇔ { to_ext(x_int) is undefined }
 *   - `SEC_ORPHAN_EXT` marks an internal value as having no matching external value
 *     SEC_ORPHAN_EXT(x_ext) ⇔ { to_int(x_ext) is undefined }
 */
template <typename InternalType, typename ExternalType, typename Tag = void>
struct SecureEnumConverter {
    using Internal = InternalType;
    using External = ExternalType;
    using Converter = SecureEnumConverter<Internal, External, Tag>;

    static SEC_OPTIONAL_NS::optional<Internal> toInternalOpt(External);
    static SEC_OPTIONAL_NS::optional<External> toExternalOpt(Internal);

    static const std::set<Internal>& convertibleInternalValues();
    static const std::set<External>& convertibleExternalValues();

    static Internal toInternalOrThrow(External external) {
        const auto& internalOpt = toInternalOpt(external);

        if (!internalOpt) {
          throw std::invalid_argument("Invalid external enum value");
        }

        return *internalOpt;
    }

    static External toExternalOrThrow(Internal internal) {
        const auto& externalOpt = toExternalOpt(internal);

        if (!externalOpt) {
          throw std::invalid_argument("Invalid internal enum value");
        }

        return *externalOpt;
    }

};

namespace priv {
    template <bool toExternal, bool toInternal, typename Converter>
    struct OneDirectionConverter;

    template <typename Converter> // To external
    struct OneDirectionConverter<true, false, Converter> {
        using Input = typename Converter::Internal;
        using Output = typename Converter::External;

        static SEC_OPTIONAL_NS::optional<Output> convertOpt(Input input)
        { return Converter::toExternalOpt(input); }

        static Output convertOrThrow(Input input)
        { return Converter::toExternalOrThrow(input); }

        static const std::set<Output>& convertibleValues()
        { return Converter::convertibleExternalValues(); }
    };

    template <typename Converter> // To internal
    struct OneDirectionConverter<false, true, Converter> {
        using Input = typename Converter::External;
        using Output = typename Converter::Internal;

        static SEC_OPTIONAL_NS::optional<Output> convertOpt(Input input)
        { return Converter::toInternalOpt(input); }

        static Output convertOrThrow(Input input)
        { return Converter::toInternalOrThrow(input); }

        static const std::set<Output>& convertibleValues()
        { return Converter::convertibleInternalValues(); }
    };
};

template <typename InternalTag, typename InternalType, typename ExternalTag, typename ExternalType, typename Tag = void>
struct TaggedEnumConverter: public SecureEnumConverter<InternalType, ExternalType, Tag> {

    template <typename DirectionTag, typename ODC = priv::OneDirectionConverter<
        std::is_same<DirectionTag, ExternalTag>::value,
        std::is_same<DirectionTag, InternalTag>::value,
        SecureEnumConverter<InternalType, ExternalType, Tag>
    > >
    static SEC_OPTIONAL_NS::optional<typename ODC::Output> convertOpt(typename ODC::Input input) {
        return ODC::convertOpt(input);
    }

    template <typename DirectionTag, typename ODC = priv::OneDirectionConverter<
        std::is_same<DirectionTag, ExternalTag>::value,
        std::is_same<DirectionTag, InternalTag>::value,
        SecureEnumConverter<InternalType, ExternalType, Tag>
    > >
    static typename ODC::Output convertOrThrow(typename ODC::Input input) {
        return ODC::convertOrThrow(input);
    }

    template <typename DirectionTag, typename ODC = priv::OneDirectionConverter<
        std::is_same<DirectionTag, ExternalTag>::value,
        std::is_same<DirectionTag, InternalTag>::value,
        SecureEnumConverter<InternalType, ExternalType, Tag>
    > >
    static std::set<typename ODC::Output> convertibleValues() {
        return ODC::convertibleValues();
    }

};

} // namespace lguim

#endif // LGUIM_SECURE_ENUM_CONVERTER
