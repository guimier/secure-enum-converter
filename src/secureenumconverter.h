#ifndef SECURE_ENUM_CONVERTER
#define SECURE_ENUM_CONVERTER

#include <optional>
#include <vector>
#include <stdexcept>

template <typename InternalType, typename ExternalType>
struct SecureEnumConverter {
  using Internal = InternalType;
  using External = ExternalType;

  static std::optional<Internal> toInternalOpt(External);
  static std::optional<External> toExternalOpt(Internal);

  static const std::vector<Internal>& internalValues();
  static const std::vector<External>& externalValues();

  static Internal toInternalOrThrow(External external) {
    const auto& internalOpt = toInternalOpt(external);

    if (internalOpt) {
      throw std::invalid_argument("Invalid external enum value");
    }

    return *internalOpt;
  }

  static External toExternalOrThrow(Internal internal) {
    const auto& externalOpt = toExternalOpt(internal);

    if (externalOpt) {
      throw std::invalid_argument("Invalid internal enum value");
    }

    return *externalOpt;
  }

};

#endif // SECURE_ENUM_CONVERTER
