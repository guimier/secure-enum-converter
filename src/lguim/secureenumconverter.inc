#include "secureenumconverter.h"

#ifdef SEC_EQUIV
    #error "SEC_EQUIV defined before including secureenumconverter.inc"
#endif

#ifdef SEC_PROJ_I2E
    #error "SEC_PROJ_I2E defined before including secureenumconverter.inc"
#endif

#ifdef SEC_PROJ_E2I
    #error "SEC_PROJ_E2I defined before including secureenumconverter.inc"
#endif

#ifdef SEC_ORPHAN_INT
    #error "SEC_ORPHAN_INT defined before including secureenumconverter.inc"
#endif

#ifdef SEC_ORPHAN_EXT
    #error "SEC_ORPHAN_EXT defined before including secureenumconverter.inc"
#endif

#ifndef SEC_TYPE
    #error "SEC_TYPE not defined"
#endif

#ifndef SEC_MAPPING
    #error "SEC_MAPPING not defined"
#endif

#pragma GCC diagnostic push
#pragma GCC diagnostic error "-Wswitch"

namespace lguim {

#ifndef SEC_NO_SWITCH_EXTERNAL
template <>
auto SEC_TYPE::Converter::toInternalOpt(External external) -> SEC_OPTIONAL_NS::optional<Internal>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) case EXT_VAL: return INT_VAL;
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL)
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL) case EXT_VAL: return INT_VAL;
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL) case EXT_VAL: return SEC_OPTIONAL_NS::nullopt;

    switch (external) {
        SEC_MAPPING
    }

    // This is unreachable if SEC_MAPPING is properly defined (which should be
    // the case if the user has no warning).
    return SEC_OPTIONAL_NS::nullopt;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}
#else // ifndef SEC_NO_SWITCH_EXTERNAL
template <>
auto SEC_TYPE::Converter::toInternalOpt(External external) -> SEC_OPTIONAL_NS::optional<Internal>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) if (external == EXT_VAL) { return INT_VAL; } else
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL)
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL) if (external == EXT_VAL) { return INT_VAL; } else
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL) if (external == EXT_VAL) { return SEC_OPTIONAL_NS::nullopt; } else

    SEC_MAPPING {
        // This is unreachable if SEC_MAPPING is properly defined.
        return SEC_OPTIONAL_NS::nullopt;
    }

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}
#endif // ifndef SEC_NO_SWITCH_EXTERNAL

#ifndef SEC_NO_SWITCH_INTERNAL
template <>
auto SEC_TYPE::Converter::toExternalOpt(Internal internal) -> SEC_OPTIONAL_NS::optional<External>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) case INT_VAL: return EXT_VAL;
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL) case INT_VAL: return EXT_VAL;
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL)
    #define SEC_ORPHAN_INT(INT_VAL) case INT_VAL: return SEC_OPTIONAL_NS::nullopt;
    #define SEC_ORPHAN_EXT(EXT_VAL)

    switch (internal) {
        SEC_MAPPING
    }

    // This is unreachable if SEC_MAPPING is properly defined (which should be
    // the case if the user has no warning).
    return SEC_OPTIONAL_NS::nullopt;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}
#else // ifndef SEC_NO_SWITCH_INTERNAL
template <>
auto SEC_TYPE::Converter::toExternalOpt(Internal internal) -> SEC_OPTIONAL_NS::optional<External>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) if (internal == INT_VAL) { return EXT_VAL; } else
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL) if (internal == INT_VAL) { return EXT_VAL; } else
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL)
    #define SEC_ORPHAN_INT(INT_VAL) if (internal == INT_VAL) { return SEC_OPTIONAL_NS::optional; } else
    #define SEC_ORPHAN_EXT(EXT_VAL)

    SEC_MAPPING {
        // This is unreachable if SEC_MAPPING is properly defined.
        return SEC_OPTIONAL_NS::nullopt;
    }

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}
#endif // ifndef SEC_NO_SWITCH_INTERNAL

template <>
auto SEC_TYPE::Converter::convertibleInternalValues() -> const std::set<Internal>&
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) INT_VAL,
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL) INT_VAL,
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL)
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL)

    static const std::set<Internal> values {
        SEC_MAPPING
    };

    return values;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

template <>
auto SEC_TYPE::Converter::convertibleExternalValues() -> const std::set<External>&
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) EXT_VAL,
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL)
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL) EXT_VAL,
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL)

    static const std::set<External> values {
        SEC_MAPPING
    };

    return values;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

} // namespace lguim

#pragma GCC diagnostic pop

#undef SEC_TYPE
#undef SEC_MAPPING
#undef SEC_NO_SWITCH_EXTERNAL
#undef SEC_NO_SWITCH_INTERNAL
