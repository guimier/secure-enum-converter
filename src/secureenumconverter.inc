#include "secureenumconverter.h"

#ifdef SEC_EQUIV
    #error "SEC_EQUIV defined before including secureenumconverter.inc"
#endif

#ifdef SEC_PROJ_I2E
    #error "SEC_PROJ_I2E defined before including secureenumconverter.inc"
#endif

#ifdef SEC_PROJ_E2I
    #error "SEC_PROJ_E2I defined before including secureenumconverter.inc"
#endif

#ifdef SEC_ORPHAN_INT
    #error "SEC_ORPHAN_INT defined before including secureenumconverter.inc"
#endif

#ifdef SEC_ORPHAN_EXT
    #error "SEC_ORPHAN_EXT defined before including secureenumconverter.inc"
#endif

#ifndef SEC_INTERNAL
    #error "SEC_INTERNAL not defined"
#endif

#ifndef SEC_EXTERNAL
    #error "SEC_EXTERNAL not defined"
#endif

#ifndef SEC_MAPPING
    #error "SEC_MAPPING not defined"
#endif

#pragma GCC diagnostic push
#pragma GCC diagnostic error "-Wswitch"

template <>
auto SecureEnumConverter<SEC_INTERNAL, SEC_EXTERNAL>::toInternalOpt(External external) -> std::optional<Internal>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) case EXT_VAL: return INT_VAL;
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL)
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL) case EXT_VAL: return INT_VAL;
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL) case EXT_VAL: return std::nullopt;

    switch (external) {
        SEC_MAPPING
    }

    // This is unreachable if SEC_MAPPING is properly defined (which should be
    // the case if the user has no warning).
    return std::nullopt;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

template <>
auto SecureEnumConverter<SEC_INTERNAL, SEC_EXTERNAL>::toExternalOpt(Internal internal) -> std::optional<External>
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) case INT_VAL: return EXT_VAL;
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL) case INT_VAL: return EXT_VAL;
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL)
    #define SEC_ORPHAN_INT(INT_VAL) case INT_VAL: return std::nullopt;
    #define SEC_ORPHAN_EXT(EXT_VAL)

    switch (internal) {
        SEC_MAPPING
    }

    // This is unreachable if SEC_MAPPING is properly defined (which should be
    // the case if the user has no warning).
    return std::nullopt;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

template <>
auto SecureEnumConverter<SEC_INTERNAL, SEC_EXTERNAL>::convertibleInternalValues() -> const std::set<Internal>&
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) INT_VAL,
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL) INT_VAL,
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL)
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL)

    static const std::set<Internal> values {
        SEC_MAPPING
    };

    return values;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

template <>
auto SecureEnumConverter<SEC_INTERNAL, SEC_EXTERNAL>::convertibleExternalValues() -> const std::set<External>&
{
    #define SEC_EQUIV(INT_VAL, EXT_VAL) EXT_VAL,
    #define SEC_PROJ_I2E(INT_VAL, EXT_VAL)
    #define SEC_PROJ_E2I(INT_VAL, EXT_VAL) EXT_VAL,
    #define SEC_ORPHAN_INT(INT_VAL)
    #define SEC_ORPHAN_EXT(EXT_VAL)

    static const std::set<External> values {
        SEC_MAPPING
    };

    return values;

    #undef SEC_ORPHAN_INT
    #undef SEC_ORPHAN_EXT
    #undef SEC_PROJ_I2E
    #undef SEC_PROJ_E2I
    #undef SEC_EQUIV
}

#pragma GCC diagnostic pop

#undef SEC_INTERNAL
#undef SEC_EXTERNAL
#undef SEC_MAPPING
