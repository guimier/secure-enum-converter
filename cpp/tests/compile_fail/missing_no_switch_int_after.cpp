#include <iostream>
#include <string>

#include "lguim/secureenumconverter.h"

enum class A { A };
enum class B { B };

using SUT1 = lguim::SecureEnumConverter<std::string, A>;
using SUT2 = lguim::SecureEnumConverter<std::string, B>;

int main () {}

#define SEC_TYPE SUT1
#define SEC_NO_SWITCH_INTERNAL
#define SEC_MAPPING \
    SEC_EQUIV("A", A::A)
#include "lguim/secureenumconverter.inc"

#define SEC_TYPE SUT2
#define SEC_MAPPING \
    SEC_EQUIV("B", B::B)
#include "lguim/secureenumconverter.inc"
