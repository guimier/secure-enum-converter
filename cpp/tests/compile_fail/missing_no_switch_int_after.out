In file included from tests/compile_fail/missing_no_switch_int_after.cpp:23:
src/lguim/secureenumconverter.inc: In static member function 'static std::optional<_Up> lguim::SecureEnumConverter<InternalType, ExternalType, Tag>::toExternalOpt(lguim::SecureEnumConverter<InternalType, ExternalType, Tag>::Internal) [with InternalType = std::__cxx11::basic_string<char>; ExternalType = B; Tag = void; lguim::SecureEnumConverter<InternalType, ExternalType, Tag>::Internal = std::__cxx11::basic_string<char>]':
src/lguim/secureenumconverter.inc:105:21: error: switch quantity not an integer
     switch (internal) {
                     ^
compilation terminated due to -Wfatal-errors.
